hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -7,247 +7,69 @@ package codegen
 
 import (
 	"context"
-	"fmt"
 
 	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/codegen/storage"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
 )
 
 // createStorageTypes returns a pipeline stage that creates dedicated storage types for each resource and nested object.
 // Storage versions are created for *all* API versions to allow users of older versions of the operator to easily
 // upgrade. This is of course a bit odd for the first release, but defining the approach from day one is useful.
-func createStorageTypes() PipelineStage {
+func createStorageTypes(idFactory astmodel.IdentifierFactory) PipelineStage {
 	return MakePipelineStage(
 		"createStorage",
 		"Create storage versions of CRD types",
 		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
 
-			storageTypes := make(astmodel.Types)
-			visitor := makeStorageTypesVisitor(types)
-			vc := makeStorageTypesVisitorContext()
-			var errs []error
-			for _, d := range types {
-				d := d
+			// Create a factory for each group (aka service) and divvy up the types
+			factories := make(map[string]*storage.StorageTypeFactory)
+			for name, def := range types {
 
-				if astmodel.ARMFlag.IsOn(d.Type()) {
-					// Skip ARM definitions, we don't need to create storage variants of those
+				ref, ok := name.PackageReference.AsLocalPackage()
+				if !ok {
+					// Skip definitions from non-local packages
+					// (should never happen)
 					continue
 				}
 
-				if _, ok := types.ResolveEnumDefinition(&d); ok {
-					// Skip Enum definitions as we use the base type for storage
+				factory, ok := factories[ref.Group()]
+				if !ok {
+					klog.V(3).Infof("Creating storage factory for %s", ref.Group())
+					factory = storage.NewStorageTypeFactory(ref.Group(), idFactory)
+					factories[ref.Group()] = factory
+				}
+
+				isArm := astmodel.ARMFlag.IsOn(def.Type())
+				if isArm {
```

Comment:
minor: probably can drop the `isArm` here, as it's not adding a lot of clarity.

If you keep it, it should be `isARM`.